{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e0a37a34-c091-488f-b6cf-290f64728090",
   "metadata": {},
   "source": [
    "## genome spot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "22b521b4-1ac2-4a4a-8124-77f770f9c07a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>oxygen</th>\n",
       "      <th>ph_max</th>\n",
       "      <th>ph_min</th>\n",
       "      <th>ph_optimum</th>\n",
       "      <th>salinity_max</th>\n",
       "      <th>salinity_min</th>\n",
       "      <th>salinity_optimum</th>\n",
       "      <th>temperature_max</th>\n",
       "      <th>temperature_min</th>\n",
       "      <th>temperature_optimum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>not tolerant</td>\n",
       "      <td>8.986309626449918</td>\n",
       "      <td>5.185028854107618</td>\n",
       "      <td>6.697489401360779</td>\n",
       "      <td>4.384593114378452</td>\n",
       "      <td>0</td>\n",
       "      <td>2.884735866415467</td>\n",
       "      <td>51.93504398348311</td>\n",
       "      <td>35.292992726330255</td>\n",
       "      <td>44.83760079640768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>not tolerant</td>\n",
       "      <td>8.394126703097054</td>\n",
       "      <td>5.517740753819724</td>\n",
       "      <td>6.682644242805267</td>\n",
       "      <td>5.3769576499020255</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>73.63443560907696</td>\n",
       "      <td>44.69383883659042</td>\n",
       "      <td>67.8920121656165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>not tolerant</td>\n",
       "      <td>8.204266073484693</td>\n",
       "      <td>4.734636137575809</td>\n",
       "      <td>6.476705995017433</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>91.20769413444867</td>\n",
       "      <td>73.4988172143595</td>\n",
       "      <td>86.60778627725867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>not tolerant</td>\n",
       "      <td>9.047247843724913</td>\n",
       "      <td>5.373528231259606</td>\n",
       "      <td>6.923415464233607</td>\n",
       "      <td>9.53042418889688</td>\n",
       "      <td>0</td>\n",
       "      <td>2.43275856665094</td>\n",
       "      <td>50.309493452094245</td>\n",
       "      <td>25.424505906619643</td>\n",
       "      <td>45.308049681186404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>tolerant</td>\n",
       "      <td>9.695084118787928</td>\n",
       "      <td>6.028665727235012</td>\n",
       "      <td>7.722688012206665</td>\n",
       "      <td>25.656326155832364</td>\n",
       "      <td>9.384982545542034</td>\n",
       "      <td>15.977666419179668</td>\n",
       "      <td>56.57626924597822</td>\n",
       "      <td>29.58442446319874</td>\n",
       "      <td>46.352929453610784</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession        oxygen             ph_max             ph_min  \\\n",
       "0  GCA_903884505  not tolerant  8.986309626449918  5.185028854107618   \n",
       "1  GCA_029259185  not tolerant  8.394126703097054  5.517740753819724   \n",
       "2  GCA_000145985  not tolerant  8.204266073484693  4.734636137575809   \n",
       "3  GCA_016927425  not tolerant  9.047247843724913  5.373528231259606   \n",
       "4  GCF_029338575      tolerant  9.695084118787928  6.028665727235012   \n",
       "\n",
       "          ph_optimum        salinity_max       salinity_min  \\\n",
       "0  6.697489401360779   4.384593114378452                  0   \n",
       "1  6.682644242805267  5.3769576499020255                  0   \n",
       "2  6.476705995017433                   0                  0   \n",
       "3  6.923415464233607    9.53042418889688                  0   \n",
       "4  7.722688012206665  25.656326155832364  9.384982545542034   \n",
       "\n",
       "     salinity_optimum     temperature_max     temperature_min  \\\n",
       "0   2.884735866415467   51.93504398348311  35.292992726330255   \n",
       "1                   0   73.63443560907696   44.69383883659042   \n",
       "2                   0   91.20769413444867    73.4988172143595   \n",
       "3    2.43275856665094  50.309493452094245  25.424505906619643   \n",
       "4  15.977666419179668   56.57626924597822   29.58442446319874   \n",
       "\n",
       "  temperature_optimum  \n",
       "0   44.83760079640768  \n",
       "1    67.8920121656165  \n",
       "2   86.60778627725867  \n",
       "3  45.308049681186404  \n",
       "4  46.352929453610784  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from glob import glob\n",
    "\n",
    "#path to file\n",
    "input_dir = \"genomespot\" \n",
    "output_file = \"genomespot_combined.tsv\"\n",
    "\n",
    "files = glob(os.path.join(input_dir, \"*.genomespot.predictions.tsv\"))\n",
    "\n",
    "records = []\n",
    "for file in files:\n",
    "    accession = os.path.basename(file).split(\".\")[0]  # strips file name, gives only accession\n",
    "    df = pd.read_csv(file, sep=\"\\t\")\n",
    "    # convert the 'value' column into a dictionary keyed by 'target'\n",
    "    data = df.set_index(\"target\")[\"value\"].to_dict()\n",
    "    data[\"accession\"] = accession\n",
    "    records.append(data)\n",
    "\n",
    "# combine into a single df\n",
    "combined_df = pd.DataFrame(records)\n",
    "\n",
    "# reorder columns (accession first)\n",
    "combined_df = combined_df.set_index('accession').reset_index()\n",
    "\n",
    "combined_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "cbf5d8bb-9976-48df-9798-64fa6e8d2051",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save file as tsv\n",
    "combined_df.to_csv(\"genomespot_all.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e810977a-f5ab-49f4-b8d5-f626641b3d56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.83760079640768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.8920121656165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.60778627725867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308049681186404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929453610784</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession temperature_optimum\n",
       "0  GCA_903884505   44.83760079640768\n",
       "1  GCA_029259185    67.8920121656165\n",
       "2  GCA_000145985   86.60778627725867\n",
       "3  GCA_016927425  45.308049681186404\n",
       "4  GCF_029338575  46.352929453610784"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#extract OGT \n",
    "\n",
    "columns = [\"accession\", \"temperature_optimum\"]\n",
    "spot_df = combined_df[columns] \n",
    "spot_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7da3980a-0ffa-4016-acb0-d7fe6da26cec",
   "metadata": {},
   "source": [
    "## genome size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "a5941903-af39-49d9-9d2c-9dff67d9847f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>Present Markers</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Redundancy</th>\n",
       "      <th>Weighted completeness</th>\n",
       "      <th>Weighted redundancy</th>\n",
       "      <th>Contigs</th>\n",
       "      <th>N50</th>\n",
       "      <th>L50</th>\n",
       "      <th>N90</th>\n",
       "      <th>L90</th>\n",
       "      <th>CDs</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_0005138551_ASM51385v1_genomic</td>\n",
       "      <td>1307099</td>\n",
       "      <td>45.16</td>\n",
       "      <td>126</td>\n",
       "      <td>0.9618</td>\n",
       "      <td>1.0079</td>\n",
       "      <td>0.9651</td>\n",
       "      <td>1.0125</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1386</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_0388711751_ASM3887117v1_genomic</td>\n",
       "      <td>1717258</td>\n",
       "      <td>41.66</td>\n",
       "      <td>117</td>\n",
       "      <td>0.8931</td>\n",
       "      <td>1.0256</td>\n",
       "      <td>0.8529</td>\n",
       "      <td>1.0441</td>\n",
       "      <td>80</td>\n",
       "      <td>29394</td>\n",
       "      <td>19</td>\n",
       "      <td>10833</td>\n",
       "      <td>57</td>\n",
       "      <td>1849</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_0232277151_ASM2322771v1_genomic</td>\n",
       "      <td>2359936</td>\n",
       "      <td>33.39</td>\n",
       "      <td>128</td>\n",
       "      <td>0.9771</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9676</td>\n",
       "      <td>1.0127</td>\n",
       "      <td>65</td>\n",
       "      <td>140569</td>\n",
       "      <td>6</td>\n",
       "      <td>43709</td>\n",
       "      <td>15</td>\n",
       "      <td>2348</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_0282797051_ASM2827970v1_genomic</td>\n",
       "      <td>1895518</td>\n",
       "      <td>46.75</td>\n",
       "      <td>112</td>\n",
       "      <td>0.8550</td>\n",
       "      <td>1.0268</td>\n",
       "      <td>0.7823</td>\n",
       "      <td>1.0503</td>\n",
       "      <td>88</td>\n",
       "      <td>145438</td>\n",
       "      <td>4</td>\n",
       "      <td>34574</td>\n",
       "      <td>15</td>\n",
       "      <td>2080</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_0001452951_ASM14529v1_genomic</td>\n",
       "      <td>1639135</td>\n",
       "      <td>48.64</td>\n",
       "      <td>129</td>\n",
       "      <td>0.9847</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9809</td>\n",
       "      <td>1.0132</td>\n",
       "      <td>2</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1750</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Name   Length  GC-content  Present Markers  \\\n",
       "0    GCF_0005138551_ASM51385v1_genomic  1307099       45.16              126   \n",
       "1  GCA_0388711751_ASM3887117v1_genomic  1717258       41.66              117   \n",
       "2  GCF_0232277151_ASM2322771v1_genomic  2359936       33.39              128   \n",
       "3  GCA_0282797051_ASM2827970v1_genomic  1895518       46.75              112   \n",
       "4    GCF_0001452951_ASM14529v1_genomic  1639135       48.64              129   \n",
       "\n",
       "   Completeness  Redundancy  Weighted completeness  Weighted redundancy  \\\n",
       "0        0.9618      1.0079                 0.9651               1.0125   \n",
       "1        0.8931      1.0256                 0.8529               1.0441   \n",
       "2        0.9771      1.0078                 0.9676               1.0127   \n",
       "3        0.8550      1.0268                 0.7823               1.0503   \n",
       "4        0.9847      1.0078                 0.9809               1.0132   \n",
       "\n",
       "   Contigs      N50  L50      N90  L90   CDs  Unnamed: 14  \n",
       "0        1  1307099    1  1307099    1  1386          NaN  \n",
       "1       80    29394   19    10833   57  1849          NaN  \n",
       "2       65   140569    6    43709   15  2348          NaN  \n",
       "3       88   145438    4    34574   15  2080          NaN  \n",
       "4        2  1634695    1  1634695    1  1750          NaN  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# corrected genome size = assembly length / completeness\n",
    "\n",
    "import pandas as pd\n",
    "mi = pd.read_csv(\"/Users/tamannatandon/Desktop/summer_project/micomplete_out/micomplete_results.tsv\", sep=\"\\t\", comment=\"#\")\n",
    "\n",
    "mi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "9de8f569-8365-4e10-9d08-4d6e9e759c7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>Present Markers</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Redundancy</th>\n",
       "      <th>Weighted completeness</th>\n",
       "      <th>Weighted redundancy</th>\n",
       "      <th>Contigs</th>\n",
       "      <th>N50</th>\n",
       "      <th>L50</th>\n",
       "      <th>N90</th>\n",
       "      <th>L90</th>\n",
       "      <th>CDs</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>genome_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_0005138551_ASM51385v1_genomic</td>\n",
       "      <td>1307099</td>\n",
       "      <td>45.16</td>\n",
       "      <td>126</td>\n",
       "      <td>0.9618</td>\n",
       "      <td>1.0079</td>\n",
       "      <td>0.9651</td>\n",
       "      <td>1.0125</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.354366e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_0388711751_ASM3887117v1_genomic</td>\n",
       "      <td>1717258</td>\n",
       "      <td>41.66</td>\n",
       "      <td>117</td>\n",
       "      <td>0.8931</td>\n",
       "      <td>1.0256</td>\n",
       "      <td>0.8529</td>\n",
       "      <td>1.0441</td>\n",
       "      <td>80</td>\n",
       "      <td>29394</td>\n",
       "      <td>19</td>\n",
       "      <td>10833</td>\n",
       "      <td>57</td>\n",
       "      <td>1849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.013434e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_0232277151_ASM2322771v1_genomic</td>\n",
       "      <td>2359936</td>\n",
       "      <td>33.39</td>\n",
       "      <td>128</td>\n",
       "      <td>0.9771</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9676</td>\n",
       "      <td>1.0127</td>\n",
       "      <td>65</td>\n",
       "      <td>140569</td>\n",
       "      <td>6</td>\n",
       "      <td>43709</td>\n",
       "      <td>15</td>\n",
       "      <td>2348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.438958e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_0282797051_ASM2827970v1_genomic</td>\n",
       "      <td>1895518</td>\n",
       "      <td>46.75</td>\n",
       "      <td>112</td>\n",
       "      <td>0.8550</td>\n",
       "      <td>1.0268</td>\n",
       "      <td>0.7823</td>\n",
       "      <td>1.0503</td>\n",
       "      <td>88</td>\n",
       "      <td>145438</td>\n",
       "      <td>4</td>\n",
       "      <td>34574</td>\n",
       "      <td>15</td>\n",
       "      <td>2080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.423007e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_0001452951_ASM14529v1_genomic</td>\n",
       "      <td>1639135</td>\n",
       "      <td>48.64</td>\n",
       "      <td>129</td>\n",
       "      <td>0.9847</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9809</td>\n",
       "      <td>1.0132</td>\n",
       "      <td>2</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.671052e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Name   Length  GC-content  Present Markers  \\\n",
       "0    GCF_0005138551_ASM51385v1_genomic  1307099       45.16              126   \n",
       "1  GCA_0388711751_ASM3887117v1_genomic  1717258       41.66              117   \n",
       "2  GCF_0232277151_ASM2322771v1_genomic  2359936       33.39              128   \n",
       "3  GCA_0282797051_ASM2827970v1_genomic  1895518       46.75              112   \n",
       "4    GCF_0001452951_ASM14529v1_genomic  1639135       48.64              129   \n",
       "\n",
       "   Completeness  Redundancy  Weighted completeness  Weighted redundancy  \\\n",
       "0        0.9618      1.0079                 0.9651               1.0125   \n",
       "1        0.8931      1.0256                 0.8529               1.0441   \n",
       "2        0.9771      1.0078                 0.9676               1.0127   \n",
       "3        0.8550      1.0268                 0.7823               1.0503   \n",
       "4        0.9847      1.0078                 0.9809               1.0132   \n",
       "\n",
       "   Contigs      N50  L50      N90  L90   CDs  Unnamed: 14   genome_size  \n",
       "0        1  1307099    1  1307099    1  1386          NaN  1.354366e+06  \n",
       "1       80    29394   19    10833   57  1849          NaN  2.013434e+06  \n",
       "2       65   140569    6    43709   15  2348          NaN  2.438958e+06  \n",
       "3       88   145438    4    34574   15  2080          NaN  2.423007e+06  \n",
       "4        2  1634695    1  1634695    1  1750          NaN  1.671052e+06  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mi[\"genome_size\"] = mi[\"Length\"] / mi[\"Weighted completeness\"] \n",
    "\n",
    "mi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "810dfd5d-af25-48d8-b7da-0419fa90ccf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>Present Markers</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Redundancy</th>\n",
       "      <th>Weighted completeness</th>\n",
       "      <th>Weighted redundancy</th>\n",
       "      <th>Contigs</th>\n",
       "      <th>N50</th>\n",
       "      <th>L50</th>\n",
       "      <th>N90</th>\n",
       "      <th>L90</th>\n",
       "      <th>CDs</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_0005138551_ASM51385v1_genomic</td>\n",
       "      <td>1307099</td>\n",
       "      <td>45.16</td>\n",
       "      <td>126</td>\n",
       "      <td>0.9618</td>\n",
       "      <td>1.0079</td>\n",
       "      <td>0.9651</td>\n",
       "      <td>1.0125</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.354366e+06</td>\n",
       "      <td>GCF_000513855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_0388711751_ASM3887117v1_genomic</td>\n",
       "      <td>1717258</td>\n",
       "      <td>41.66</td>\n",
       "      <td>117</td>\n",
       "      <td>0.8931</td>\n",
       "      <td>1.0256</td>\n",
       "      <td>0.8529</td>\n",
       "      <td>1.0441</td>\n",
       "      <td>80</td>\n",
       "      <td>29394</td>\n",
       "      <td>19</td>\n",
       "      <td>10833</td>\n",
       "      <td>57</td>\n",
       "      <td>1849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.013434e+06</td>\n",
       "      <td>GCA_038871175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_0232277151_ASM2322771v1_genomic</td>\n",
       "      <td>2359936</td>\n",
       "      <td>33.39</td>\n",
       "      <td>128</td>\n",
       "      <td>0.9771</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9676</td>\n",
       "      <td>1.0127</td>\n",
       "      <td>65</td>\n",
       "      <td>140569</td>\n",
       "      <td>6</td>\n",
       "      <td>43709</td>\n",
       "      <td>15</td>\n",
       "      <td>2348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.438958e+06</td>\n",
       "      <td>GCF_023227715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_0282797051_ASM2827970v1_genomic</td>\n",
       "      <td>1895518</td>\n",
       "      <td>46.75</td>\n",
       "      <td>112</td>\n",
       "      <td>0.8550</td>\n",
       "      <td>1.0268</td>\n",
       "      <td>0.7823</td>\n",
       "      <td>1.0503</td>\n",
       "      <td>88</td>\n",
       "      <td>145438</td>\n",
       "      <td>4</td>\n",
       "      <td>34574</td>\n",
       "      <td>15</td>\n",
       "      <td>2080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.423007e+06</td>\n",
       "      <td>GCA_028279705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_0001452951_ASM14529v1_genomic</td>\n",
       "      <td>1639135</td>\n",
       "      <td>48.64</td>\n",
       "      <td>129</td>\n",
       "      <td>0.9847</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9809</td>\n",
       "      <td>1.0132</td>\n",
       "      <td>2</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.671052e+06</td>\n",
       "      <td>GCF_000145295</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Name   Length  GC-content  Present Markers  \\\n",
       "0    GCF_0005138551_ASM51385v1_genomic  1307099       45.16              126   \n",
       "1  GCA_0388711751_ASM3887117v1_genomic  1717258       41.66              117   \n",
       "2  GCF_0232277151_ASM2322771v1_genomic  2359936       33.39              128   \n",
       "3  GCA_0282797051_ASM2827970v1_genomic  1895518       46.75              112   \n",
       "4    GCF_0001452951_ASM14529v1_genomic  1639135       48.64              129   \n",
       "\n",
       "   Completeness  Redundancy  Weighted completeness  Weighted redundancy  \\\n",
       "0        0.9618      1.0079                 0.9651               1.0125   \n",
       "1        0.8931      1.0256                 0.8529               1.0441   \n",
       "2        0.9771      1.0078                 0.9676               1.0127   \n",
       "3        0.8550      1.0268                 0.7823               1.0503   \n",
       "4        0.9847      1.0078                 0.9809               1.0132   \n",
       "\n",
       "   Contigs      N50  L50      N90  L90   CDs  Unnamed: 14   genome_size  \\\n",
       "0        1  1307099    1  1307099    1  1386          NaN  1.354366e+06   \n",
       "1       80    29394   19    10833   57  1849          NaN  2.013434e+06   \n",
       "2       65   140569    6    43709   15  2348          NaN  2.438958e+06   \n",
       "3       88   145438    4    34574   15  2080          NaN  2.423007e+06   \n",
       "4        2  1634695    1  1634695    1  1750          NaN  1.671052e+06   \n",
       "\n",
       "       accession  \n",
       "0  GCF_000513855  \n",
       "1  GCA_038871175  \n",
       "2  GCF_023227715  \n",
       "3  GCA_028279705  \n",
       "4  GCF_000145295  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mi[\"accession\"] = mi[\"Name\"].str.extract(r\"(GC[AF]_\\d{9})\")\n",
    "mi.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "724c6cb5-8371-41db-940e-1057415d0624",
   "metadata": {},
   "source": [
    "## gene density"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "4f003c39-a5d6-4f53-998d-b16e842dc2a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Length</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>Present Markers</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Redundancy</th>\n",
       "      <th>Weighted completeness</th>\n",
       "      <th>Weighted redundancy</th>\n",
       "      <th>Contigs</th>\n",
       "      <th>N50</th>\n",
       "      <th>L50</th>\n",
       "      <th>N90</th>\n",
       "      <th>L90</th>\n",
       "      <th>CDs</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>accession</th>\n",
       "      <th>gene_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_0005138551_ASM51385v1_genomic</td>\n",
       "      <td>1307099</td>\n",
       "      <td>45.16</td>\n",
       "      <td>126</td>\n",
       "      <td>0.9618</td>\n",
       "      <td>1.0079</td>\n",
       "      <td>0.9651</td>\n",
       "      <td>1.0125</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1307099</td>\n",
       "      <td>1</td>\n",
       "      <td>1386</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.354366e+06</td>\n",
       "      <td>GCF_000513855</td>\n",
       "      <td>1060.363446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_0388711751_ASM3887117v1_genomic</td>\n",
       "      <td>1717258</td>\n",
       "      <td>41.66</td>\n",
       "      <td>117</td>\n",
       "      <td>0.8931</td>\n",
       "      <td>1.0256</td>\n",
       "      <td>0.8529</td>\n",
       "      <td>1.0441</td>\n",
       "      <td>80</td>\n",
       "      <td>29394</td>\n",
       "      <td>19</td>\n",
       "      <td>10833</td>\n",
       "      <td>57</td>\n",
       "      <td>1849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.013434e+06</td>\n",
       "      <td>GCA_038871175</td>\n",
       "      <td>1076.716486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_0232277151_ASM2322771v1_genomic</td>\n",
       "      <td>2359936</td>\n",
       "      <td>33.39</td>\n",
       "      <td>128</td>\n",
       "      <td>0.9771</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9676</td>\n",
       "      <td>1.0127</td>\n",
       "      <td>65</td>\n",
       "      <td>140569</td>\n",
       "      <td>6</td>\n",
       "      <td>43709</td>\n",
       "      <td>15</td>\n",
       "      <td>2348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.438958e+06</td>\n",
       "      <td>GCF_023227715</td>\n",
       "      <td>994.942236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_0282797051_ASM2827970v1_genomic</td>\n",
       "      <td>1895518</td>\n",
       "      <td>46.75</td>\n",
       "      <td>112</td>\n",
       "      <td>0.8550</td>\n",
       "      <td>1.0268</td>\n",
       "      <td>0.7823</td>\n",
       "      <td>1.0503</td>\n",
       "      <td>88</td>\n",
       "      <td>145438</td>\n",
       "      <td>4</td>\n",
       "      <td>34574</td>\n",
       "      <td>15</td>\n",
       "      <td>2080</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.423007e+06</td>\n",
       "      <td>GCA_028279705</td>\n",
       "      <td>1097.325375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_0001452951_ASM14529v1_genomic</td>\n",
       "      <td>1639135</td>\n",
       "      <td>48.64</td>\n",
       "      <td>129</td>\n",
       "      <td>0.9847</td>\n",
       "      <td>1.0078</td>\n",
       "      <td>0.9809</td>\n",
       "      <td>1.0132</td>\n",
       "      <td>2</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1634695</td>\n",
       "      <td>1</td>\n",
       "      <td>1750</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.671052e+06</td>\n",
       "      <td>GCF_000145295</td>\n",
       "      <td>1067.636284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Name   Length  GC-content  Present Markers  \\\n",
       "0    GCF_0005138551_ASM51385v1_genomic  1307099       45.16              126   \n",
       "1  GCA_0388711751_ASM3887117v1_genomic  1717258       41.66              117   \n",
       "2  GCF_0232277151_ASM2322771v1_genomic  2359936       33.39              128   \n",
       "3  GCA_0282797051_ASM2827970v1_genomic  1895518       46.75              112   \n",
       "4    GCF_0001452951_ASM14529v1_genomic  1639135       48.64              129   \n",
       "\n",
       "   Completeness  Redundancy  Weighted completeness  Weighted redundancy  \\\n",
       "0        0.9618      1.0079                 0.9651               1.0125   \n",
       "1        0.8931      1.0256                 0.8529               1.0441   \n",
       "2        0.9771      1.0078                 0.9676               1.0127   \n",
       "3        0.8550      1.0268                 0.7823               1.0503   \n",
       "4        0.9847      1.0078                 0.9809               1.0132   \n",
       "\n",
       "   Contigs      N50  L50      N90  L90   CDs  Unnamed: 14   genome_size  \\\n",
       "0        1  1307099    1  1307099    1  1386          NaN  1.354366e+06   \n",
       "1       80    29394   19    10833   57  1849          NaN  2.013434e+06   \n",
       "2       65   140569    6    43709   15  2348          NaN  2.438958e+06   \n",
       "3       88   145438    4    34574   15  2080          NaN  2.423007e+06   \n",
       "4        2  1634695    1  1634695    1  1750          NaN  1.671052e+06   \n",
       "\n",
       "       accession  gene_density  \n",
       "0  GCF_000513855   1060.363446  \n",
       "1  GCA_038871175   1076.716486  \n",
       "2  GCF_023227715    994.942236  \n",
       "3  GCA_028279705   1097.325375  \n",
       "4  GCF_000145295   1067.636284  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mi[\"gene_density\"] = mi[\"CDs\"] / (mi[\"Length\"] / 1000000) \n",
    "mi.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0a0a3cc-355f-496c-a324-ff14e581a63a",
   "metadata": {},
   "source": [
    "## summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "109eb046-65bb-4ad1-8b2f-21224fea6ff3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_000513855</td>\n",
       "      <td>1.354366e+06</td>\n",
       "      <td>1060.363446</td>\n",
       "      <td>45.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_038871175</td>\n",
       "      <td>2.013434e+06</td>\n",
       "      <td>1076.716486</td>\n",
       "      <td>41.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_023227715</td>\n",
       "      <td>2.438958e+06</td>\n",
       "      <td>994.942236</td>\n",
       "      <td>33.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_028279705</td>\n",
       "      <td>2.423007e+06</td>\n",
       "      <td>1097.325375</td>\n",
       "      <td>46.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_000145295</td>\n",
       "      <td>1.671052e+06</td>\n",
       "      <td>1067.636284</td>\n",
       "      <td>48.64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession   genome_size  gene_density  GC-content\n",
       "0  GCF_000513855  1.354366e+06   1060.363446       45.16\n",
       "1  GCA_038871175  2.013434e+06   1076.716486       41.66\n",
       "2  GCF_023227715  2.438958e+06    994.942236       33.39\n",
       "3  GCA_028279705  2.423007e+06   1097.325375       46.75\n",
       "4  GCF_000145295  1.671052e+06   1067.636284       48.64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#extract relevant columns \n",
    "\n",
    "columns = [\"accession\", \"genome_size\", \"gene_density\", \"GC-content\"]\n",
    "mi_df = mi[columns]\n",
    "mi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e85b2687-2194-4b59-ab2c-1b73f956ece4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.83760079640768</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.8920121656165</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.60778627725867</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308049681186404</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929453610784</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession temperature_optimum   genome_size  gene_density  GC-content\n",
       "0  GCA_903884505   44.83760079640768  1.759997e+06   1030.709851       56.76\n",
       "1  GCA_029259185    67.8920121656165  1.350382e+06   1126.294884       51.96\n",
       "2  GCA_000145985   86.60778627725867  1.942986e+06   1065.591729       35.69\n",
       "3  GCA_016927425  45.308049681186404  2.358715e+06    979.862129       56.05\n",
       "4  GCF_029338575  46.352929453610784  4.428875e+06    999.456175       67.92"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#combine by accession\n",
    "\n",
    "summary = spot_df.merge(mi_df, on = \"accession\", how=\"left\")\n",
    "summary.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "19aa3929-cc91-4180-b70a-606db48a8399",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save to file\n",
    "summary.to_csv(\"summary.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab96a013-4306-4b96-88d4-066098fcaaae",
   "metadata": {},
   "source": [
    "## protein features\n",
    "- protein length\n",
    "- fraction of charged aa (DEKR)\n",
    "- fraction of polar amino acids (STNQ)\n",
    "- fraction of ILVWYGERKP i.e. thermostability group\n",
    "- hydropathicity (GRAVY)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "99d93ef0-fde4-4fae-baf8-fb6cfa9bd140",
   "metadata": {},
   "outputs": [],
   "source": [
    "#simple file parser\n",
    "\n",
    "def fasta_parser(file):\n",
    "    with open(file, \"r\") as f:\n",
    "        seq = ''\n",
    "        for line in f:\n",
    "            if line.startswith(\">\"):\n",
    "                if seq:\n",
    "                    yield seq\n",
    "                    seq = ''\n",
    "            else:\n",
    "                seq += line.strip()\n",
    "        if seq:\n",
    "            yield seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "24df9240-764d-419e-b53a-0b2390da10e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#amino acid feature calculation\n",
    "\n",
    "def compute_features(faa_file):\n",
    "    charged = \"DEKR\"\n",
    "    polar = \"STNQ\"\n",
    "    ilvwyg = \"ILVWYGERKP\"\n",
    "    hydrophobic = \"AVILMFWY\"\n",
    "    aromatic = \"FYW\"\n",
    "\n",
    "    #kyte-doolittle values\n",
    "    hydropathy = {\n",
    "        'A': 1.8, 'R': -4.5, 'N': -3.5, 'D': -3.5, 'C': 2.5,\n",
    "        'Q': -3.5, 'E': -3.5, 'G': -0.4, 'H': -3.2, 'I': 4.5,\n",
    "        'L': 3.8, 'K': -3.9, 'M': 1.9, 'F': 2.8, 'P': -1.6,\n",
    "        'S': -0.8, 'T': -0.7, 'W': -0.9, 'Y': -1.3, 'V': 4.2} \n",
    "\n",
    "    total_aa = 0\n",
    "    counts = {\n",
    "        \"charged\": 0,\n",
    "        \"polar\": 0,\n",
    "        \"hydrophobic\": 0,\n",
    "        \"aromatic\": 0,\n",
    "        \"ilvwyg\": 0,\n",
    "        \"gravy_sum\": 0.0,\n",
    "        \"protein_count\": 0\n",
    "    }\n",
    "\n",
    "    for seq in fasta_parser(faa_file):\n",
    "        seq = seq.upper()\n",
    "        if not seq:\n",
    "            continue\n",
    "        counts[\"protein_count\"] += 1\n",
    "        total_aa += len(seq)\n",
    "        for aa in seq:\n",
    "            if aa in charged: \n",
    "                counts[\"charged\"] += 1\n",
    "            if aa in polar: \n",
    "                counts[\"polar\"] += 1\n",
    "            if aa in hydrophobic: \n",
    "                counts[\"hydrophobic\"] += 1\n",
    "            if aa in aromatic:\n",
    "                counts[\"aromatic\"] += 1\n",
    "            if aa in ilvwyg:\n",
    "                counts[\"ilvwyg\"] += 1\n",
    "            if aa in hydropathy: \n",
    "                counts[\"gravy_sum\"] += hydropathy[aa]\n",
    "\n",
    "    if total_aa == 0:\n",
    "        return None\n",
    "\n",
    "    return {\"aa_total\": total_aa,\n",
    "        \"proteins\": counts[\"protein_count\"],\n",
    "        \"frac_charged\": counts[\"charged\"] / total_aa,\n",
    "        \"frac_polar\": counts[\"polar\"] / total_aa,\n",
    "        \"frac_hydrophobic\": counts[\"hydrophobic\"] / total_aa,\n",
    "        \"frac_aromatic\": counts[\"aromatic\"] / total_aa,\n",
    "        \"frac_ilvwyg\": counts[\"ilvwyg\"] / total_aa,\n",
    "        \"gravy_mean\": counts[\"gravy_sum\"] / total_aa,}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "d133c3dd-de61-47a3-ac5e-b7aac246571d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "records = []\n",
    "for faa in glob(\"prokka_out/prokka_*/*.faa\"):\n",
    "    accession = \"_\".join(os.path.basename(faa).split(\".\")[0].split(\"_\")[1:])\n",
    "    feats = compute_features(faa)\n",
    "    if feats:\n",
    "        feats[\"accession\"] = accession\n",
    "        records.append(feats)\n",
    "\n",
    "aa_df = pd.DataFrame(records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "a27be246-ae99-4dba-9496-67e62dadde9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>797456</td>\n",
       "      <td>2863</td>\n",
       "      <td>0.246355</td>\n",
       "      <td>0.190705</td>\n",
       "      <td>0.424899</td>\n",
       "      <td>0.083566</td>\n",
       "      <td>0.575569</td>\n",
       "      <td>-0.137496</td>\n",
       "      <td>GCA_030611265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>452523</td>\n",
       "      <td>1604</td>\n",
       "      <td>0.247839</td>\n",
       "      <td>0.154291</td>\n",
       "      <td>0.436809</td>\n",
       "      <td>0.073408</td>\n",
       "      <td>0.611953</td>\n",
       "      <td>-0.059902</td>\n",
       "      <td>GCF_000711905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>678559</td>\n",
       "      <td>2203</td>\n",
       "      <td>0.224437</td>\n",
       "      <td>0.171110</td>\n",
       "      <td>0.436650</td>\n",
       "      <td>0.080867</td>\n",
       "      <td>0.577023</td>\n",
       "      <td>-0.024972</td>\n",
       "      <td>GCF_018687785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>464455</td>\n",
       "      <td>1720</td>\n",
       "      <td>0.233990</td>\n",
       "      <td>0.187226</td>\n",
       "      <td>0.437080</td>\n",
       "      <td>0.082544</td>\n",
       "      <td>0.591493</td>\n",
       "      <td>-0.071790</td>\n",
       "      <td>GCA_016871995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>425647</td>\n",
       "      <td>1502</td>\n",
       "      <td>0.246253</td>\n",
       "      <td>0.182870</td>\n",
       "      <td>0.440454</td>\n",
       "      <td>0.089100</td>\n",
       "      <td>0.620850</td>\n",
       "      <td>-0.052428</td>\n",
       "      <td>GCA_001940665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   aa_total  proteins  frac_charged  frac_polar  frac_hydrophobic  \\\n",
       "0    797456      2863      0.246355    0.190705          0.424899   \n",
       "1    452523      1604      0.247839    0.154291          0.436809   \n",
       "2    678559      2203      0.224437    0.171110          0.436650   \n",
       "3    464455      1720      0.233990    0.187226          0.437080   \n",
       "4    425647      1502      0.246253    0.182870          0.440454   \n",
       "\n",
       "   frac_aromatic  frac_ilvwyg  gravy_mean      accession  \n",
       "0       0.083566     0.575569   -0.137496  GCA_030611265  \n",
       "1       0.073408     0.611953   -0.059902  GCF_000711905  \n",
       "2       0.080867     0.577023   -0.024972  GCF_018687785  \n",
       "3       0.082544     0.591493   -0.071790  GCA_016871995  \n",
       "4       0.089100     0.620850   -0.052428  GCA_001940665  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "e0eba507-b4e1-47bd-ae07-8f5338bc95b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>accession_clean_x</th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "      <th>accession_clean_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.83760079640768</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>460626.0</td>\n",
       "      <td>1530.0</td>\n",
       "      <td>0.237657</td>\n",
       "      <td>0.175387</td>\n",
       "      <td>0.428979</td>\n",
       "      <td>0.082095</td>\n",
       "      <td>0.579288</td>\n",
       "      <td>-0.131588</td>\n",
       "      <td>GCA_903884505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.8920121656165</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>373399.0</td>\n",
       "      <td>1378.0</td>\n",
       "      <td>0.266768</td>\n",
       "      <td>0.148048</td>\n",
       "      <td>0.423065</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.641070</td>\n",
       "      <td>-0.140035</td>\n",
       "      <td>GCA_029259185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.60778627725867</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>544497.0</td>\n",
       "      <td>1994.0</td>\n",
       "      <td>0.241572</td>\n",
       "      <td>0.168431</td>\n",
       "      <td>0.464671</td>\n",
       "      <td>0.094364</td>\n",
       "      <td>0.635449</td>\n",
       "      <td>0.052572</td>\n",
       "      <td>GCA_000145985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308049681186404</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>663102.0</td>\n",
       "      <td>2210.0</td>\n",
       "      <td>0.238681</td>\n",
       "      <td>0.165441</td>\n",
       "      <td>0.432558</td>\n",
       "      <td>0.075074</td>\n",
       "      <td>0.591933</td>\n",
       "      <td>-0.078561</td>\n",
       "      <td>GCA_016927425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929453610784</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>1199013.0</td>\n",
       "      <td>4285.0</td>\n",
       "      <td>0.252032</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.428929</td>\n",
       "      <td>0.072433</td>\n",
       "      <td>0.574337</td>\n",
       "      <td>-0.153448</td>\n",
       "      <td>GCF_029338575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession temperature_optimum   genome_size  gene_density  GC-content  \\\n",
       "0  GCA_903884505   44.83760079640768  1.759997e+06   1030.709851       56.76   \n",
       "1  GCA_029259185    67.8920121656165  1.350382e+06   1126.294884       51.96   \n",
       "2  GCA_000145985   86.60778627725867  1.942986e+06   1065.591729       35.69   \n",
       "3  GCA_016927425  45.308049681186404  2.358715e+06    979.862129       56.05   \n",
       "4  GCF_029338575  46.352929453610784  4.428875e+06    999.456175       67.92   \n",
       "\n",
       "  accession_clean_x   aa_total  proteins  frac_charged  frac_polar  \\\n",
       "0     GCA_903884505   460626.0    1530.0      0.237657    0.175387   \n",
       "1     GCA_029259185   373399.0    1378.0      0.266768    0.148048   \n",
       "2     GCA_000145985   544497.0    1994.0      0.241572    0.168431   \n",
       "3     GCA_016927425   663102.0    2210.0      0.238681    0.165441   \n",
       "4     GCF_029338575  1199013.0    4285.0      0.252032    0.152640   \n",
       "\n",
       "   frac_hydrophobic  frac_aromatic  frac_ilvwyg  gravy_mean accession_clean_y  \n",
       "0          0.428979       0.082095     0.579288   -0.131588     GCA_903884505  \n",
       "1          0.423065       0.077228     0.641070   -0.140035     GCA_029259185  \n",
       "2          0.464671       0.094364     0.635449    0.052572     GCA_000145985  \n",
       "3          0.432558       0.075074     0.591933   -0.078561     GCA_016927425  \n",
       "4          0.428929       0.072433     0.574337   -0.153448     GCF_029338575  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#combine by accession\n",
    "\n",
    "merged = summary.merge(aa_df, on = \"accession\", how=\"left\") \n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "944e7ce0-c3a3-4448-92d9-4d49248c81b0",
   "metadata": {},
   "source": [
    "It appears as if there is some mismatch between accessions, which needs to be inspected to be fixed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "99fbd041-b7b7-4384-87f0-3549a0aa7c2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing in aa_df:\n",
      "GCA_002009975\n",
      "GCA_002505765\n",
      "GCA_003166335\n",
      "GCA_003584625\n",
      "GCA_004211975\n",
      "GCA_013572355\n",
      "GCA_014859785\n",
      "GCA_017656505\n",
      "GCA_019084405\n",
      "GCA_021159385\n",
      "GCA_025061095\n",
      "GCA_027036775\n",
      "GCA_030055515\n",
      "GCA_038741495\n",
      "GCF_000014945\n",
      "GCF_000063445\n",
      "GCF_000191585\n",
      "GCF_000194625\n",
      "GCF_000404225\n",
      "GCF_000455345\n",
      "GCF_000455365\n",
      "GCF_001484685\n",
      "GCF_012027325\n",
      "GCF_017873855\n",
      "GCF_017874455\n",
      "GCF_018128925\n",
      "GCF_023227715\n",
      "GCF_030464345\n",
      "\n",
      "Missing in main_df:\n",
      "A_GCA_003166335\n",
      "A_GCA_003584625\n",
      "A_GCA_013572355\n",
      "A_GCA_014859785\n",
      "A_GCA_017656505\n",
      "A_GCA_030055515\n",
      "A_GCF_000063445\n",
      "A_GCF_000404225\n",
      "A_GCF_000455345\n",
      "A_GCF_001484685\n",
      "A_GCF_018128925\n",
      "A_GCF_030464345\n",
      "B_GCA_025061095\n",
      "B_GCA_027036775\n",
      "B_GCF_000014945\n",
      "B_GCF_000191585\n",
      "B_GCF_000455365\n",
      "B_GCF_012027325\n",
      "B_GCF_017873855\n",
      "C_GCA_002009975\n",
      "C_GCA_021159385\n",
      "C_GCA_038741495\n",
      "C_GCF_000194625\n",
      "C_GCF_017874455\n",
      "D_GCA_002505765\n",
      "D_GCA_004211975\n",
      "E_GCF_023227715\n",
      "F_GCA_019084405\n"
     ]
    }
   ],
   "source": [
    "summary[\"accession\"] = summary[\"accession\"].apply(strip_version)\n",
    "aa_df[\"accession\"] = aa_df[\"accession\"].apply(strip_version)\n",
    "\n",
    "main_accessions = set(summary[\"accession\"])\n",
    "aa_accessions = set(aa_df[\"accession\"])\n",
    "\n",
    "# In main_df but missing from aa_df\n",
    "missing_in_aa = main_accessions - aa_accessions\n",
    "\n",
    "# In aa_df but missing from main_df\n",
    "missing_in_main = aa_accessions - main_accessions\n",
    "\n",
    "print(\"Missing in aa_df:\")\n",
    "for acc in sorted(missing_in_aa):\n",
    "    print(acc)\n",
    "\n",
    "print(\"\\nMissing in main_df:\")\n",
    "for acc in sorted(missing_in_main):\n",
    "    print(acc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "5f3c5238-8738-414a-b676-f29edced47bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>accession_clean_x</th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "      <th>accession_clean_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.83760079640768</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>460626</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.237657</td>\n",
       "      <td>0.175387</td>\n",
       "      <td>0.428979</td>\n",
       "      <td>0.082095</td>\n",
       "      <td>0.579288</td>\n",
       "      <td>-0.131588</td>\n",
       "      <td>GCA_903884505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.8920121656165</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>373399</td>\n",
       "      <td>1378</td>\n",
       "      <td>0.266768</td>\n",
       "      <td>0.148048</td>\n",
       "      <td>0.423065</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.641070</td>\n",
       "      <td>-0.140035</td>\n",
       "      <td>GCA_029259185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.60778627725867</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>544497</td>\n",
       "      <td>1994</td>\n",
       "      <td>0.241572</td>\n",
       "      <td>0.168431</td>\n",
       "      <td>0.464671</td>\n",
       "      <td>0.094364</td>\n",
       "      <td>0.635449</td>\n",
       "      <td>0.052572</td>\n",
       "      <td>GCA_000145985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308049681186404</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>663102</td>\n",
       "      <td>2210</td>\n",
       "      <td>0.238681</td>\n",
       "      <td>0.165441</td>\n",
       "      <td>0.432558</td>\n",
       "      <td>0.075074</td>\n",
       "      <td>0.591933</td>\n",
       "      <td>-0.078561</td>\n",
       "      <td>GCA_016927425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929453610784</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>1199013</td>\n",
       "      <td>4285</td>\n",
       "      <td>0.252032</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.428929</td>\n",
       "      <td>0.072433</td>\n",
       "      <td>0.574337</td>\n",
       "      <td>-0.153448</td>\n",
       "      <td>GCF_029338575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession temperature_optimum   genome_size  gene_density  GC-content  \\\n",
       "0  GCA_903884505   44.83760079640768  1.759997e+06   1030.709851       56.76   \n",
       "1  GCA_029259185    67.8920121656165  1.350382e+06   1126.294884       51.96   \n",
       "2  GCA_000145985   86.60778627725867  1.942986e+06   1065.591729       35.69   \n",
       "3  GCA_016927425  45.308049681186404  2.358715e+06    979.862129       56.05   \n",
       "4  GCF_029338575  46.352929453610784  4.428875e+06    999.456175       67.92   \n",
       "\n",
       "  accession_clean_x  aa_total  proteins  frac_charged  frac_polar  \\\n",
       "0     GCA_903884505    460626      1530      0.237657    0.175387   \n",
       "1     GCA_029259185    373399      1378      0.266768    0.148048   \n",
       "2     GCA_000145985    544497      1994      0.241572    0.168431   \n",
       "3     GCA_016927425    663102      2210      0.238681    0.165441   \n",
       "4     GCF_029338575   1199013      4285      0.252032    0.152640   \n",
       "\n",
       "   frac_hydrophobic  frac_aromatic  frac_ilvwyg  gravy_mean accession_clean_y  \n",
       "0          0.428979       0.082095     0.579288   -0.131588     GCA_903884505  \n",
       "1          0.423065       0.077228     0.641070   -0.140035     GCA_029259185  \n",
       "2          0.464671       0.094364     0.635449    0.052572     GCA_000145985  \n",
       "3          0.432558       0.075074     0.591933   -0.078561     GCA_016927425  \n",
       "4          0.428929       0.072433     0.574337   -0.153448     GCF_029338575  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#strip and match the accessions\n",
    "\n",
    "summary[\"accession\"] = summary[\"accession\"].str.extract(r\"(GCA_\\d+|GCF_\\d+)\")\n",
    "aa_df[\"accession\"] = aa_df[\"accession\"].str.extract(r\"(GCA_\\d+|GCF_\\d+)\")\n",
    "merged = pd.merge(summary, aa_df, on=\"accession\", how=\"inner\")\n",
    "\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "3df1a198-43b5-4726-b8e6-88f09b036a27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.83760079640768</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "      <td>460626</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.237657</td>\n",
       "      <td>0.175387</td>\n",
       "      <td>0.428979</td>\n",
       "      <td>0.082095</td>\n",
       "      <td>0.579288</td>\n",
       "      <td>-0.131588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.8920121656165</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "      <td>373399</td>\n",
       "      <td>1378</td>\n",
       "      <td>0.266768</td>\n",
       "      <td>0.148048</td>\n",
       "      <td>0.423065</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.641070</td>\n",
       "      <td>-0.140035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.60778627725867</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "      <td>544497</td>\n",
       "      <td>1994</td>\n",
       "      <td>0.241572</td>\n",
       "      <td>0.168431</td>\n",
       "      <td>0.464671</td>\n",
       "      <td>0.094364</td>\n",
       "      <td>0.635449</td>\n",
       "      <td>0.052572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308049681186404</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "      <td>663102</td>\n",
       "      <td>2210</td>\n",
       "      <td>0.238681</td>\n",
       "      <td>0.165441</td>\n",
       "      <td>0.432558</td>\n",
       "      <td>0.075074</td>\n",
       "      <td>0.591933</td>\n",
       "      <td>-0.078561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929453610784</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "      <td>1199013</td>\n",
       "      <td>4285</td>\n",
       "      <td>0.252032</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.428929</td>\n",
       "      <td>0.072433</td>\n",
       "      <td>0.574337</td>\n",
       "      <td>-0.153448</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession temperature_optimum   genome_size  gene_density  GC-content  \\\n",
       "0  GCA_903884505   44.83760079640768  1.759997e+06   1030.709851       56.76   \n",
       "1  GCA_029259185    67.8920121656165  1.350382e+06   1126.294884       51.96   \n",
       "2  GCA_000145985   86.60778627725867  1.942986e+06   1065.591729       35.69   \n",
       "3  GCA_016927425  45.308049681186404  2.358715e+06    979.862129       56.05   \n",
       "4  GCF_029338575  46.352929453610784  4.428875e+06    999.456175       67.92   \n",
       "\n",
       "   aa_total  proteins  frac_charged  frac_polar  frac_hydrophobic  \\\n",
       "0    460626      1530      0.237657    0.175387          0.428979   \n",
       "1    373399      1378      0.266768    0.148048          0.423065   \n",
       "2    544497      1994      0.241572    0.168431          0.464671   \n",
       "3    663102      2210      0.238681    0.165441          0.432558   \n",
       "4   1199013      4285      0.252032    0.152640          0.428929   \n",
       "\n",
       "   frac_aromatic  frac_ilvwyg  gravy_mean  \n",
       "0       0.082095     0.579288   -0.131588  \n",
       "1       0.077228     0.641070   -0.140035  \n",
       "2       0.094364     0.635449    0.052572  \n",
       "3       0.075074     0.591933   -0.078561  \n",
       "4       0.072433     0.574337   -0.153448  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#rename one to accession and drop\n",
    "merged[\"accession\"] = merged[\"accession_clean_x\"] \n",
    "merged = merged.drop(columns=[\"accession_clean_x\", \"accession_clean_y\"])\n",
    "\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "1c7d4d9b-a52e-4096-b1c7-d4d320df4dea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save as file\n",
    "merged.to_csv(\"summary.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c560bc1-b853-4879-8aef-04ca5372b1ac",
   "metadata": {},
   "source": [
    "## **tRNA features**\n",
    "- GC%\n",
    "- average tRNA length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ca401a00-8776-4835-8f0b-3593e8f126fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>tRNA_GC_mean</th>\n",
       "      <th>avg_tRNA_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_020854815</td>\n",
       "      <td>61.549369</td>\n",
       "      <td>78.978723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029982165</td>\n",
       "      <td>61.483544</td>\n",
       "      <td>81.010753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_029887765</td>\n",
       "      <td>63.403735</td>\n",
       "      <td>82.769784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_018814355</td>\n",
       "      <td>60.239999</td>\n",
       "      <td>84.005464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_015520265</td>\n",
       "      <td>71.878877</td>\n",
       "      <td>86.924107</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  tRNA_GC_mean  avg_tRNA_length\n",
       "0  GCA_020854815     61.549369        78.978723\n",
       "1  GCA_029982165     61.483544        81.010753\n",
       "2  GCA_029887765     63.403735        82.769784\n",
       "3  GCA_018814355     60.239999        84.005464\n",
       "4  GCA_015520265     71.878877        86.924107"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "import pandas as pd\n",
    "\n",
    "#find all files\n",
    "fna_files = glob.glob(\"trna_out/*_trna.fna\")\n",
    "\n",
    "rows = []\n",
    "lengths = []\n",
    "\n",
    "for fna_file in fna_files:\n",
    "    accession = fna_file.split(\"/\")[-1].replace(\"_trna.fna\", \"\")\n",
    "\n",
    "    gc_values = []\n",
    "    with open(fna_file) as f:\n",
    "        seq = \"\"\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if seq:\n",
    "                    gc = (seq.count(\"G\") + seq.count(\"C\")) / len(seq) * 100\n",
    "                    gc_values.append(gc)\n",
    "                    lengths.append(len(seq))\n",
    "                    seq = \"\"\n",
    "            else:\n",
    "                seq += line.strip()\n",
    "        if seq != \"\":\n",
    "            lengths.append(len(seq))\n",
    "                \n",
    "        #processing last sequence\n",
    "        if seq:\n",
    "            gc = (seq.count(\"G\") + seq.count(\"C\")) / len(seq) * 100\n",
    "            gc_values.append(gc)\n",
    "\n",
    "    avg_length = sum(lengths) / len(lengths) if lengths else 0\n",
    "    #calculating mean GC%\n",
    "    if gc_values:\n",
    "        mean_gc = sum(gc_values) / len(gc_values)\n",
    "    else:\n",
    "         mean_gc = None \n",
    "    \n",
    "    rows.append({\"accession\": accession, \"tRNA_GC_mean\": mean_gc, \"avg_tRNA_length\": avg_length})\n",
    "\n",
    "#save results as tsv\n",
    "df = pd.DataFrame(rows)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4c9e37d5-5adb-41ca-a9ba-80e3af5b61b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.837601</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "      <td>460626</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.237657</td>\n",
       "      <td>0.175387</td>\n",
       "      <td>0.428979</td>\n",
       "      <td>0.082095</td>\n",
       "      <td>0.579288</td>\n",
       "      <td>-0.131588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.892012</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "      <td>373399</td>\n",
       "      <td>1378</td>\n",
       "      <td>0.266768</td>\n",
       "      <td>0.148048</td>\n",
       "      <td>0.423065</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.641070</td>\n",
       "      <td>-0.140035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.607786</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "      <td>544497</td>\n",
       "      <td>1994</td>\n",
       "      <td>0.241572</td>\n",
       "      <td>0.168431</td>\n",
       "      <td>0.464671</td>\n",
       "      <td>0.094364</td>\n",
       "      <td>0.635449</td>\n",
       "      <td>0.052572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308050</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "      <td>663102</td>\n",
       "      <td>2210</td>\n",
       "      <td>0.238681</td>\n",
       "      <td>0.165441</td>\n",
       "      <td>0.432558</td>\n",
       "      <td>0.075074</td>\n",
       "      <td>0.591933</td>\n",
       "      <td>-0.078561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "      <td>1199013</td>\n",
       "      <td>4285</td>\n",
       "      <td>0.252032</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.428929</td>\n",
       "      <td>0.072433</td>\n",
       "      <td>0.574337</td>\n",
       "      <td>-0.153448</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  temperature_optimum   genome_size  gene_density  GC-content  \\\n",
       "0  GCA_903884505            44.837601  1.759997e+06   1030.709851       56.76   \n",
       "1  GCA_029259185            67.892012  1.350382e+06   1126.294884       51.96   \n",
       "2  GCA_000145985            86.607786  1.942986e+06   1065.591729       35.69   \n",
       "3  GCA_016927425            45.308050  2.358715e+06    979.862129       56.05   \n",
       "4  GCF_029338575            46.352929  4.428875e+06    999.456175       67.92   \n",
       "\n",
       "   aa_total  proteins  frac_charged  frac_polar  frac_hydrophobic  \\\n",
       "0    460626      1530      0.237657    0.175387          0.428979   \n",
       "1    373399      1378      0.266768    0.148048          0.423065   \n",
       "2    544497      1994      0.241572    0.168431          0.464671   \n",
       "3    663102      2210      0.238681    0.165441          0.432558   \n",
       "4   1199013      4285      0.252032    0.152640          0.428929   \n",
       "\n",
       "   frac_aromatic  frac_ilvwyg  gravy_mean  \n",
       "0       0.082095     0.579288   -0.131588  \n",
       "1       0.077228     0.641070   -0.140035  \n",
       "2       0.094364     0.635449    0.052572  \n",
       "3       0.075074     0.591933   -0.078561  \n",
       "4       0.072433     0.574337   -0.153448  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary = pd.read_csv(\"summary.tsv\", sep=\"\\t\")\n",
    "summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a7c99acc-c7b8-43ae-bf5f-459858c0ccb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "      <th>tRNA_GC_mean</th>\n",
       "      <th>avg_tRNA_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.837601</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "      <td>460626</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.237657</td>\n",
       "      <td>0.175387</td>\n",
       "      <td>0.428979</td>\n",
       "      <td>0.082095</td>\n",
       "      <td>0.579288</td>\n",
       "      <td>-0.131588</td>\n",
       "      <td>59.950459</td>\n",
       "      <td>90.292916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.892012</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "      <td>373399</td>\n",
       "      <td>1378</td>\n",
       "      <td>0.266768</td>\n",
       "      <td>0.148048</td>\n",
       "      <td>0.423065</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.641070</td>\n",
       "      <td>-0.140035</td>\n",
       "      <td>64.962643</td>\n",
       "      <td>89.959412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.607786</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "      <td>544497</td>\n",
       "      <td>1994</td>\n",
       "      <td>0.241572</td>\n",
       "      <td>0.168431</td>\n",
       "      <td>0.464671</td>\n",
       "      <td>0.094364</td>\n",
       "      <td>0.635449</td>\n",
       "      <td>0.052572</td>\n",
       "      <td>69.203992</td>\n",
       "      <td>88.351724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308050</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "      <td>663102</td>\n",
       "      <td>2210</td>\n",
       "      <td>0.238681</td>\n",
       "      <td>0.165441</td>\n",
       "      <td>0.432558</td>\n",
       "      <td>0.075074</td>\n",
       "      <td>0.591933</td>\n",
       "      <td>-0.078561</td>\n",
       "      <td>61.710844</td>\n",
       "      <td>90.439763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "      <td>1199013</td>\n",
       "      <td>4285</td>\n",
       "      <td>0.252032</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.428929</td>\n",
       "      <td>0.072433</td>\n",
       "      <td>0.574337</td>\n",
       "      <td>-0.153448</td>\n",
       "      <td>61.729627</td>\n",
       "      <td>88.052546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  temperature_optimum   genome_size  gene_density  GC-content  \\\n",
       "0  GCA_903884505            44.837601  1.759997e+06   1030.709851       56.76   \n",
       "1  GCA_029259185            67.892012  1.350382e+06   1126.294884       51.96   \n",
       "2  GCA_000145985            86.607786  1.942986e+06   1065.591729       35.69   \n",
       "3  GCA_016927425            45.308050  2.358715e+06    979.862129       56.05   \n",
       "4  GCF_029338575            46.352929  4.428875e+06    999.456175       67.92   \n",
       "\n",
       "   aa_total  proteins  frac_charged  frac_polar  frac_hydrophobic  \\\n",
       "0    460626      1530      0.237657    0.175387          0.428979   \n",
       "1    373399      1378      0.266768    0.148048          0.423065   \n",
       "2    544497      1994      0.241572    0.168431          0.464671   \n",
       "3    663102      2210      0.238681    0.165441          0.432558   \n",
       "4   1199013      4285      0.252032    0.152640          0.428929   \n",
       "\n",
       "   frac_aromatic  frac_ilvwyg  gravy_mean  tRNA_GC_mean  avg_tRNA_length  \n",
       "0       0.082095     0.579288   -0.131588     59.950459        90.292916  \n",
       "1       0.077228     0.641070   -0.140035     64.962643        89.959412  \n",
       "2       0.094364     0.635449    0.052572     69.203992        88.351724  \n",
       "3       0.075074     0.591933   -0.078561     61.710844        90.439763  \n",
       "4       0.072433     0.574337   -0.153448     61.729627        88.052546  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.merge(summary, df, on=\"accession\", how=\"inner\")\n",
    "\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ca158124-fa91-499a-a361-e083254fac6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save as file\n",
    "merged.to_csv(\"summary.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56d98aae-c73e-4a6b-8f8b-def0cba11edd",
   "metadata": {},
   "source": [
    "## **codonw features**\n",
    "- GC3s\n",
    "- Gravy\n",
    "- aromo\n",
    "- Nc "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b49f9a06-900f-456a-a474-86f5350dde83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>T3s</th>\n",
       "      <th>C3s</th>\n",
       "      <th>A3s</th>\n",
       "      <th>G3s</th>\n",
       "      <th>CAI</th>\n",
       "      <th>CBI</th>\n",
       "      <th>Fop</th>\n",
       "      <th>Nc</th>\n",
       "      <th>GC3s</th>\n",
       "      <th>GC</th>\n",
       "      <th>L_sym</th>\n",
       "      <th>L_aa</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Aromo</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Average_of_genes</td>\n",
       "      <td>0.0764</td>\n",
       "      <td>0.5981</td>\n",
       "      <td>0.1059</td>\n",
       "      <td>0.4366</td>\n",
       "      <td>0.244</td>\n",
       "      <td>0.195</td>\n",
       "      <td>0.531</td>\n",
       "      <td>36.97</td>\n",
       "      <td>0.849</td>\n",
       "      <td>0.664</td>\n",
       "      <td>1275432</td>\n",
       "      <td>1313095</td>\n",
       "      <td>-0.244439</td>\n",
       "      <td>0.073193</td>\n",
       "      <td>GCF_023238205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Average_of_genes</td>\n",
       "      <td>0.2393</td>\n",
       "      <td>0.3610</td>\n",
       "      <td>0.2916</td>\n",
       "      <td>0.3512</td>\n",
       "      <td>0.180</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.426</td>\n",
       "      <td>54.42</td>\n",
       "      <td>0.563</td>\n",
       "      <td>0.501</td>\n",
       "      <td>460714</td>\n",
       "      <td>477945</td>\n",
       "      <td>-0.072605</td>\n",
       "      <td>0.082934</td>\n",
       "      <td>GCA_038852285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Average_of_genes</td>\n",
       "      <td>0.3606</td>\n",
       "      <td>0.2477</td>\n",
       "      <td>0.4140</td>\n",
       "      <td>0.2868</td>\n",
       "      <td>0.193</td>\n",
       "      <td>-0.024</td>\n",
       "      <td>0.407</td>\n",
       "      <td>47.92</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.395</td>\n",
       "      <td>374724</td>\n",
       "      <td>386831</td>\n",
       "      <td>-0.234110</td>\n",
       "      <td>0.093312</td>\n",
       "      <td>GCA_018303625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Average_of_genes</td>\n",
       "      <td>0.3754</td>\n",
       "      <td>0.2306</td>\n",
       "      <td>0.3994</td>\n",
       "      <td>0.2419</td>\n",
       "      <td>0.170</td>\n",
       "      <td>-0.027</td>\n",
       "      <td>0.391</td>\n",
       "      <td>52.45</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.421</td>\n",
       "      <td>543397</td>\n",
       "      <td>564529</td>\n",
       "      <td>-0.031203</td>\n",
       "      <td>0.094495</td>\n",
       "      <td>GCA_018396755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Average_of_genes</td>\n",
       "      <td>0.1751</td>\n",
       "      <td>0.4052</td>\n",
       "      <td>0.2041</td>\n",
       "      <td>0.4327</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.405</td>\n",
       "      <td>45.86</td>\n",
       "      <td>0.679</td>\n",
       "      <td>0.563</td>\n",
       "      <td>496116</td>\n",
       "      <td>513689</td>\n",
       "      <td>0.065056</td>\n",
       "      <td>0.080928</td>\n",
       "      <td>GCA_015521325</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              title     T3s     C3s     A3s     G3s    CAI     CBI    Fop  \\\n",
       "0  Average_of_genes  0.0764  0.5981  0.1059  0.4366  0.244   0.195  0.531   \n",
       "1  Average_of_genes  0.2393  0.3610  0.2916  0.3512  0.180   0.031  0.426   \n",
       "2  Average_of_genes  0.3606  0.2477  0.4140  0.2868  0.193  -0.024  0.407   \n",
       "3  Average_of_genes  0.3754  0.2306  0.3994  0.2419  0.170  -0.027  0.391   \n",
       "4  Average_of_genes  0.1751  0.4052  0.2041  0.4327  0.161   0.002  0.405   \n",
       "\n",
       "      Nc   GC3s     GC    L_sym     L_aa      Gravy     Aromo      accession  \n",
       "0  36.97  0.849  0.664  1275432  1313095  -0.244439  0.073193  GCF_023238205  \n",
       "1  54.42  0.563  0.501   460714   477945  -0.072605  0.082934  GCA_038852285  \n",
       "2  47.92  0.393  0.395   374724   386831  -0.234110  0.093312  GCA_018303625  \n",
       "3  52.45  0.367  0.421   543397   564529  -0.031203  0.094495  GCA_018396755  \n",
       "4  45.86  0.679  0.563   496116   513689   0.065056  0.080928  GCA_015521325  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "#collect .out files\n",
    "out_files = glob.glob(\"gc3_results/*_cds.out\")\n",
    "\n",
    "all_features = []\n",
    "\n",
    "for out_file in out_files: \n",
    "    accession = out_file.split(\"/\")[-1].replace(\"_cds.out\", \"\")\n",
    "\n",
    "    with open(out_file) as file:\n",
    "        lines = []\n",
    "        for line in file: \n",
    "            line = line.strip() \n",
    "            if line != \"\":\n",
    "                lines.append(line) \n",
    "    #column names\n",
    "    header_line = lines[0]\n",
    "    header = header_line.split()\n",
    "\n",
    "    #data values\n",
    "    value_line = lines[1]\n",
    "    values = value_line.split()\n",
    "\n",
    "    #make dictionary (column:value) \n",
    "    feature={}\n",
    "    for i in range(len(header)): \n",
    "        feature[header[i]]=values[i]\n",
    "    \n",
    "    feature['accession']=accession\n",
    "\n",
    "    #save genome's features\n",
    "    all_features.append(feature) \n",
    "\n",
    "\n",
    "#convert to dataframe\n",
    "codonw_df = pd.DataFrame(all_features) \n",
    "codonw_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7e1448cd-3fb4-4b1d-a906-cd69c16b6e87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>GC3s</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Aromo</th>\n",
       "      <th>Nc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_023238205</td>\n",
       "      <td>0.849</td>\n",
       "      <td>-0.244439</td>\n",
       "      <td>0.073193</td>\n",
       "      <td>36.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_038852285</td>\n",
       "      <td>0.563</td>\n",
       "      <td>-0.072605</td>\n",
       "      <td>0.082934</td>\n",
       "      <td>54.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_018303625</td>\n",
       "      <td>0.393</td>\n",
       "      <td>-0.234110</td>\n",
       "      <td>0.093312</td>\n",
       "      <td>47.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_018396755</td>\n",
       "      <td>0.367</td>\n",
       "      <td>-0.031203</td>\n",
       "      <td>0.094495</td>\n",
       "      <td>52.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_015521325</td>\n",
       "      <td>0.679</td>\n",
       "      <td>0.065056</td>\n",
       "      <td>0.080928</td>\n",
       "      <td>45.86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession   GC3s      Gravy     Aromo     Nc\n",
       "0  GCF_023238205  0.849  -0.244439  0.073193  36.97\n",
       "1  GCA_038852285  0.563  -0.072605  0.082934  54.42\n",
       "2  GCA_018303625  0.393  -0.234110  0.093312  47.92\n",
       "3  GCA_018396755  0.367  -0.031203  0.094495  52.45\n",
       "4  GCA_015521325  0.679   0.065056  0.080928  45.86"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#extract relevant features\n",
    "\n",
    "columns = [\"accession\", \"GC3s\", \"Gravy\", \"Aromo\", \"Nc\"]\n",
    "\n",
    "feature_df = codonw_df[columns]\n",
    "feature_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "08d5db1e-f151-4228-b5d5-c4cacb797620",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "      <th>tRNA_GC_mean</th>\n",
       "      <th>avg_tRNA_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.837601</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "      <td>460626</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.237657</td>\n",
       "      <td>0.175387</td>\n",
       "      <td>0.428979</td>\n",
       "      <td>0.082095</td>\n",
       "      <td>0.579288</td>\n",
       "      <td>-0.131588</td>\n",
       "      <td>59.950459</td>\n",
       "      <td>90.292916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.892012</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "      <td>373399</td>\n",
       "      <td>1378</td>\n",
       "      <td>0.266768</td>\n",
       "      <td>0.148048</td>\n",
       "      <td>0.423065</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.641070</td>\n",
       "      <td>-0.140035</td>\n",
       "      <td>64.962643</td>\n",
       "      <td>89.959412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.607786</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "      <td>544497</td>\n",
       "      <td>1994</td>\n",
       "      <td>0.241572</td>\n",
       "      <td>0.168431</td>\n",
       "      <td>0.464671</td>\n",
       "      <td>0.094364</td>\n",
       "      <td>0.635449</td>\n",
       "      <td>0.052572</td>\n",
       "      <td>69.203992</td>\n",
       "      <td>88.351724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308050</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "      <td>663102</td>\n",
       "      <td>2210</td>\n",
       "      <td>0.238681</td>\n",
       "      <td>0.165441</td>\n",
       "      <td>0.432558</td>\n",
       "      <td>0.075074</td>\n",
       "      <td>0.591933</td>\n",
       "      <td>-0.078561</td>\n",
       "      <td>61.710844</td>\n",
       "      <td>90.439763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "      <td>1199013</td>\n",
       "      <td>4285</td>\n",
       "      <td>0.252032</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.428929</td>\n",
       "      <td>0.072433</td>\n",
       "      <td>0.574337</td>\n",
       "      <td>-0.153448</td>\n",
       "      <td>61.729627</td>\n",
       "      <td>88.052546</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  temperature_optimum   genome_size  gene_density  GC-content  \\\n",
       "0  GCA_903884505            44.837601  1.759997e+06   1030.709851       56.76   \n",
       "1  GCA_029259185            67.892012  1.350382e+06   1126.294884       51.96   \n",
       "2  GCA_000145985            86.607786  1.942986e+06   1065.591729       35.69   \n",
       "3  GCA_016927425            45.308050  2.358715e+06    979.862129       56.05   \n",
       "4  GCF_029338575            46.352929  4.428875e+06    999.456175       67.92   \n",
       "\n",
       "   aa_total  proteins  frac_charged  frac_polar  frac_hydrophobic  \\\n",
       "0    460626      1530      0.237657    0.175387          0.428979   \n",
       "1    373399      1378      0.266768    0.148048          0.423065   \n",
       "2    544497      1994      0.241572    0.168431          0.464671   \n",
       "3    663102      2210      0.238681    0.165441          0.432558   \n",
       "4   1199013      4285      0.252032    0.152640          0.428929   \n",
       "\n",
       "   frac_aromatic  frac_ilvwyg  gravy_mean  tRNA_GC_mean  avg_tRNA_length  \n",
       "0       0.082095     0.579288   -0.131588     59.950459        90.292916  \n",
       "1       0.077228     0.641070   -0.140035     64.962643        89.959412  \n",
       "2       0.094364     0.635449    0.052572     69.203992        88.351724  \n",
       "3       0.075074     0.591933   -0.078561     61.710844        90.439763  \n",
       "4       0.072433     0.574337   -0.153448     61.729627        88.052546  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load summary.tsv\n",
    "\n",
    "summary = pd.read_csv(\"summary.tsv\", sep=\"\\t\")\n",
    "\n",
    "summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7a503e08-c3e9-480b-804a-d449ee7f5253",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>temperature_optimum</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>GC-content</th>\n",
       "      <th>aa_total</th>\n",
       "      <th>proteins</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>gravy_mean</th>\n",
       "      <th>tRNA_GC_mean</th>\n",
       "      <th>avg_tRNA_length</th>\n",
       "      <th>GC3s</th>\n",
       "      <th>Gravy</th>\n",
       "      <th>Aromo</th>\n",
       "      <th>Nc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.837601</td>\n",
       "      <td>1.759997e+06</td>\n",
       "      <td>1030.709851</td>\n",
       "      <td>56.76</td>\n",
       "      <td>460626</td>\n",
       "      <td>1530</td>\n",
       "      <td>0.237657</td>\n",
       "      <td>0.175387</td>\n",
       "      <td>0.428979</td>\n",
       "      <td>0.082095</td>\n",
       "      <td>0.579288</td>\n",
       "      <td>-0.131588</td>\n",
       "      <td>59.950459</td>\n",
       "      <td>90.292916</td>\n",
       "      <td>0.726</td>\n",
       "      <td>-0.130644</td>\n",
       "      <td>0.082074</td>\n",
       "      <td>45.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.892012</td>\n",
       "      <td>1.350382e+06</td>\n",
       "      <td>1126.294884</td>\n",
       "      <td>51.96</td>\n",
       "      <td>373399</td>\n",
       "      <td>1378</td>\n",
       "      <td>0.266768</td>\n",
       "      <td>0.148048</td>\n",
       "      <td>0.423065</td>\n",
       "      <td>0.077228</td>\n",
       "      <td>0.641070</td>\n",
       "      <td>-0.140035</td>\n",
       "      <td>64.962643</td>\n",
       "      <td>89.959412</td>\n",
       "      <td>0.594</td>\n",
       "      <td>-0.138000</td>\n",
       "      <td>0.077224</td>\n",
       "      <td>53.88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.607786</td>\n",
       "      <td>1.942986e+06</td>\n",
       "      <td>1065.591729</td>\n",
       "      <td>35.69</td>\n",
       "      <td>544497</td>\n",
       "      <td>1994</td>\n",
       "      <td>0.241572</td>\n",
       "      <td>0.168431</td>\n",
       "      <td>0.464671</td>\n",
       "      <td>0.094364</td>\n",
       "      <td>0.635449</td>\n",
       "      <td>0.052572</td>\n",
       "      <td>69.203992</td>\n",
       "      <td>88.351724</td>\n",
       "      <td>0.241</td>\n",
       "      <td>0.054511</td>\n",
       "      <td>0.094426</td>\n",
       "      <td>41.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.308050</td>\n",
       "      <td>2.358715e+06</td>\n",
       "      <td>979.862129</td>\n",
       "      <td>56.05</td>\n",
       "      <td>663102</td>\n",
       "      <td>2210</td>\n",
       "      <td>0.238681</td>\n",
       "      <td>0.165441</td>\n",
       "      <td>0.432558</td>\n",
       "      <td>0.075074</td>\n",
       "      <td>0.591933</td>\n",
       "      <td>-0.078561</td>\n",
       "      <td>61.710844</td>\n",
       "      <td>90.439763</td>\n",
       "      <td>0.686</td>\n",
       "      <td>-0.077339</td>\n",
       "      <td>0.075046</td>\n",
       "      <td>49.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.352929</td>\n",
       "      <td>4.428875e+06</td>\n",
       "      <td>999.456175</td>\n",
       "      <td>67.92</td>\n",
       "      <td>1199013</td>\n",
       "      <td>4285</td>\n",
       "      <td>0.252032</td>\n",
       "      <td>0.152640</td>\n",
       "      <td>0.428929</td>\n",
       "      <td>0.072433</td>\n",
       "      <td>0.574337</td>\n",
       "      <td>-0.153448</td>\n",
       "      <td>61.729627</td>\n",
       "      <td>88.052546</td>\n",
       "      <td>0.898</td>\n",
       "      <td>-0.152036</td>\n",
       "      <td>0.072411</td>\n",
       "      <td>33.59</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession  temperature_optimum   genome_size  gene_density  GC-content  \\\n",
       "0  GCA_903884505            44.837601  1.759997e+06   1030.709851       56.76   \n",
       "1  GCA_029259185            67.892012  1.350382e+06   1126.294884       51.96   \n",
       "2  GCA_000145985            86.607786  1.942986e+06   1065.591729       35.69   \n",
       "3  GCA_016927425            45.308050  2.358715e+06    979.862129       56.05   \n",
       "4  GCF_029338575            46.352929  4.428875e+06    999.456175       67.92   \n",
       "\n",
       "   aa_total  proteins  frac_charged  frac_polar  frac_hydrophobic  \\\n",
       "0    460626      1530      0.237657    0.175387          0.428979   \n",
       "1    373399      1378      0.266768    0.148048          0.423065   \n",
       "2    544497      1994      0.241572    0.168431          0.464671   \n",
       "3    663102      2210      0.238681    0.165441          0.432558   \n",
       "4   1199013      4285      0.252032    0.152640          0.428929   \n",
       "\n",
       "   frac_aromatic  frac_ilvwyg  gravy_mean  tRNA_GC_mean  avg_tRNA_length  \\\n",
       "0       0.082095     0.579288   -0.131588     59.950459        90.292916   \n",
       "1       0.077228     0.641070   -0.140035     64.962643        89.959412   \n",
       "2       0.094364     0.635449    0.052572     69.203992        88.351724   \n",
       "3       0.075074     0.591933   -0.078561     61.710844        90.439763   \n",
       "4       0.072433     0.574337   -0.153448     61.729627        88.052546   \n",
       "\n",
       "    GC3s      Gravy     Aromo     Nc  \n",
       "0  0.726  -0.130644  0.082074  45.67  \n",
       "1  0.594  -0.138000  0.077224  53.88  \n",
       "2  0.241   0.054511  0.094426  41.70  \n",
       "3  0.686  -0.077339  0.075046  49.23  \n",
       "4  0.898  -0.152036  0.072411  33.59  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge dfs\n",
    "\n",
    "merged = pd.merge(summary, feature_df, on=\"accession\", how=\"inner\")\n",
    "\n",
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6cd83e87-e3e0-42fe-ae76-05d5957345da",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save as file\n",
    "merged.to_csv(\"summary.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f826443-5529-4018-bd2a-fb38c16986dc",
   "metadata": {},
   "source": [
    "## additional features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5e31977c-8610-4c2d-a6a7-d89483b75bab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ogt</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>trna_gc</th>\n",
       "      <th>trna_length</th>\n",
       "      <th>gc3s</th>\n",
       "      <th>gravy</th>\n",
       "      <th>aromo</th>\n",
       "      <th>nc</th>\n",
       "      <th>ratio_charged_polar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.84</td>\n",
       "      <td>1.76</td>\n",
       "      <td>1030.71</td>\n",
       "      <td>0.238</td>\n",
       "      <td>0.175</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.579</td>\n",
       "      <td>59.95</td>\n",
       "      <td>90.29</td>\n",
       "      <td>72.6</td>\n",
       "      <td>-0.131</td>\n",
       "      <td>0.082</td>\n",
       "      <td>45.67</td>\n",
       "      <td>1.360000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.89</td>\n",
       "      <td>1.35</td>\n",
       "      <td>1126.29</td>\n",
       "      <td>0.267</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.423</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.641</td>\n",
       "      <td>64.96</td>\n",
       "      <td>89.96</td>\n",
       "      <td>59.4</td>\n",
       "      <td>-0.138</td>\n",
       "      <td>0.077</td>\n",
       "      <td>53.88</td>\n",
       "      <td>1.804054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.61</td>\n",
       "      <td>1.94</td>\n",
       "      <td>1065.59</td>\n",
       "      <td>0.242</td>\n",
       "      <td>0.168</td>\n",
       "      <td>0.465</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.635</td>\n",
       "      <td>69.20</td>\n",
       "      <td>88.35</td>\n",
       "      <td>24.1</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.094</td>\n",
       "      <td>41.70</td>\n",
       "      <td>1.440476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.31</td>\n",
       "      <td>2.36</td>\n",
       "      <td>979.86</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.592</td>\n",
       "      <td>61.71</td>\n",
       "      <td>90.44</td>\n",
       "      <td>68.6</td>\n",
       "      <td>-0.077</td>\n",
       "      <td>0.075</td>\n",
       "      <td>49.23</td>\n",
       "      <td>1.448485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.35</td>\n",
       "      <td>4.43</td>\n",
       "      <td>999.46</td>\n",
       "      <td>0.252</td>\n",
       "      <td>0.153</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.574</td>\n",
       "      <td>61.73</td>\n",
       "      <td>88.05</td>\n",
       "      <td>89.8</td>\n",
       "      <td>-0.152</td>\n",
       "      <td>0.072</td>\n",
       "      <td>33.59</td>\n",
       "      <td>1.647059</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession    ogt  genome_size  gene_density  frac_charged  frac_polar  \\\n",
       "0  GCA_903884505  44.84         1.76       1030.71         0.238       0.175   \n",
       "1  GCA_029259185  67.89         1.35       1126.29         0.267       0.148   \n",
       "2  GCA_000145985  86.61         1.94       1065.59         0.242       0.168   \n",
       "3  GCA_016927425  45.31         2.36        979.86         0.239       0.165   \n",
       "4  GCF_029338575  46.35         4.43        999.46         0.252       0.153   \n",
       "\n",
       "   frac_hydrophobic  frac_aromatic  frac_ilvwyg  trna_gc  trna_length  gc3s  \\\n",
       "0             0.429          0.082        0.579    59.95        90.29  72.6   \n",
       "1             0.423          0.077        0.641    64.96        89.96  59.4   \n",
       "2             0.465          0.094        0.635    69.20        88.35  24.1   \n",
       "3             0.433          0.075        0.592    61.71        90.44  68.6   \n",
       "4             0.429          0.072        0.574    61.73        88.05  89.8   \n",
       "\n",
       "   gravy  aromo     nc  ratio_charged_polar  \n",
       "0 -0.131  0.082  45.67             1.360000  \n",
       "1 -0.138  0.077  53.88             1.804054  \n",
       "2  0.055  0.094  41.70             1.440476  \n",
       "3 -0.077  0.075  49.23             1.448485  \n",
       "4 -0.152  0.072  33.59             1.647059  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load file\n",
    "\n",
    "import pandas as pd \n",
    "\n",
    "clean_file = pd.read_csv(\"summary_clean.tsv\", sep=\"\\t\")\n",
    "clean_file.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5b991a20-a8c9-4860-863a-9836d718a9e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ogt</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>gene_density</th>\n",
       "      <th>frac_charged</th>\n",
       "      <th>frac_polar</th>\n",
       "      <th>frac_hydrophobic</th>\n",
       "      <th>frac_aromatic</th>\n",
       "      <th>frac_ilvwyg</th>\n",
       "      <th>trna_gc</th>\n",
       "      <th>trna_length</th>\n",
       "      <th>gc3s</th>\n",
       "      <th>gravy</th>\n",
       "      <th>aromo</th>\n",
       "      <th>nc</th>\n",
       "      <th>ratio_charged_polar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_903884505</td>\n",
       "      <td>44.84</td>\n",
       "      <td>1.76</td>\n",
       "      <td>1030.71</td>\n",
       "      <td>0.238</td>\n",
       "      <td>0.175</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.579</td>\n",
       "      <td>59.95</td>\n",
       "      <td>90.29</td>\n",
       "      <td>72.6</td>\n",
       "      <td>-0.131</td>\n",
       "      <td>0.082</td>\n",
       "      <td>45.67</td>\n",
       "      <td>1.360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_029259185</td>\n",
       "      <td>67.89</td>\n",
       "      <td>1.35</td>\n",
       "      <td>1126.29</td>\n",
       "      <td>0.267</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.423</td>\n",
       "      <td>0.077</td>\n",
       "      <td>0.641</td>\n",
       "      <td>64.96</td>\n",
       "      <td>89.96</td>\n",
       "      <td>59.4</td>\n",
       "      <td>-0.138</td>\n",
       "      <td>0.077</td>\n",
       "      <td>53.88</td>\n",
       "      <td>1.804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000145985</td>\n",
       "      <td>86.61</td>\n",
       "      <td>1.94</td>\n",
       "      <td>1065.59</td>\n",
       "      <td>0.242</td>\n",
       "      <td>0.168</td>\n",
       "      <td>0.465</td>\n",
       "      <td>0.094</td>\n",
       "      <td>0.635</td>\n",
       "      <td>69.20</td>\n",
       "      <td>88.35</td>\n",
       "      <td>24.1</td>\n",
       "      <td>0.055</td>\n",
       "      <td>0.094</td>\n",
       "      <td>41.70</td>\n",
       "      <td>1.440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_016927425</td>\n",
       "      <td>45.31</td>\n",
       "      <td>2.36</td>\n",
       "      <td>979.86</td>\n",
       "      <td>0.239</td>\n",
       "      <td>0.165</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0.075</td>\n",
       "      <td>0.592</td>\n",
       "      <td>61.71</td>\n",
       "      <td>90.44</td>\n",
       "      <td>68.6</td>\n",
       "      <td>-0.077</td>\n",
       "      <td>0.075</td>\n",
       "      <td>49.23</td>\n",
       "      <td>1.448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_029338575</td>\n",
       "      <td>46.35</td>\n",
       "      <td>4.43</td>\n",
       "      <td>999.46</td>\n",
       "      <td>0.252</td>\n",
       "      <td>0.153</td>\n",
       "      <td>0.429</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.574</td>\n",
       "      <td>61.73</td>\n",
       "      <td>88.05</td>\n",
       "      <td>89.8</td>\n",
       "      <td>-0.152</td>\n",
       "      <td>0.072</td>\n",
       "      <td>33.59</td>\n",
       "      <td>1.647</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession    ogt  genome_size  gene_density  frac_charged  frac_polar  \\\n",
       "0  GCA_903884505  44.84         1.76       1030.71         0.238       0.175   \n",
       "1  GCA_029259185  67.89         1.35       1126.29         0.267       0.148   \n",
       "2  GCA_000145985  86.61         1.94       1065.59         0.242       0.168   \n",
       "3  GCA_016927425  45.31         2.36        979.86         0.239       0.165   \n",
       "4  GCF_029338575  46.35         4.43        999.46         0.252       0.153   \n",
       "\n",
       "   frac_hydrophobic  frac_aromatic  frac_ilvwyg  trna_gc  trna_length  gc3s  \\\n",
       "0             0.429          0.082        0.579    59.95        90.29  72.6   \n",
       "1             0.423          0.077        0.641    64.96        89.96  59.4   \n",
       "2             0.465          0.094        0.635    69.20        88.35  24.1   \n",
       "3             0.433          0.075        0.592    61.71        90.44  68.6   \n",
       "4             0.429          0.072        0.574    61.73        88.05  89.8   \n",
       "\n",
       "   gravy  aromo     nc  ratio_charged_polar  \n",
       "0 -0.131  0.082  45.67                1.360  \n",
       "1 -0.138  0.077  53.88                1.804  \n",
       "2  0.055  0.094  41.70                1.440  \n",
       "3 -0.077  0.075  49.23                1.448  \n",
       "4 -0.152  0.072  33.59                1.647  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#add ratio of frac_charged / frac_polar\n",
    "\n",
    "clean_file[\"ratio_charged_polar\"] = clean_file[\"frac_charged\"] / clean_file[\"frac_polar\"]\n",
    "\n",
    "#round number \n",
    "clean_file = clean_file.round({\"ratio_charged_polar\" : 3})\n",
    "\n",
    "clean_file.head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "add9ee82-2d28-4aeb-af06-50097194c0ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save as file\n",
    "clean_file.to_csv(\"summary_clean.tsv\", sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
